### PR TITLE
Multipart LC Expiration Support

### LINKED ISSUE #642: rgw/sfs: Support namespace RGW_OBJ_NS_MULTIPART bucket->list() for multipart LC expiration
### What needs to be done

LC multipart expiration uses bucket-> list with currently unsupported options. Support those. See code for TODO comments

### PR DESCRIPTION
**Pull request by irq0**
Support multipart expiration through RGW LC.

Add `RGW_OBJ_NS_MULTIPART` support to bucket listings and a special LC
path to `get_multipart_upload` to find back LCed multiparts.

Rename MP object_uuid to path_uuid to avoid confusion with object
table uuid / name.

Epic: https://github.com/aquarist-labs/s3gw/issues/256
Epic: https://github.com/aquarist-labs/s3gw/issues/216
Fixes: https://github.com/aquarist-labs/s3gw/issues/642

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

### CODE REVIEW

File: src/rgw/driver/sfs/bucket.cc [hunk 1]
```diff
@@ -385,14 +460,33 @@ std::unique_ptr<MultipartUpload> SFSBucket::get_multipart_upload(
       << "bucket::" << __func__ << ": oid: " << with_oid
       << ", upload id: " << with_upload_id << dendl;
 
-  std::string id = with_upload_id.value_or("");
-  if (id.empty()) {
-    id = bucket->gen_multipart_upload_id();
+  std::string next_oid(with_oid);
```
Comment (0xavi0) [inline src/rgw/driver/sfs/bucket.cc:463]: Question.

What happens if, for whatever reason, `try_resolve_mp_from_oid` is not successful?
If my understanding is correct, it'll call `next_upload_id = with_upload_id.value_or("");`, but `next_oid` will be already assigned to the value of `with_oid`.

Maybe this is a corner case, but I'm just curious, because in that case we return a multipart with a value that maybe is not the one we were looking for.

What's confusing to me is that in the `try_resolve_mp_from_oid` function `next_oid` is returned with the value of the multipart `object_name`, but in this particular case it would be returned with the `oid` passed from `LC`.

**Review by 0xavi0 — COMMENTED** [review]
looks good.
Just 1 question to be sure I'm understanding this correctly

File: src/rgw/driver/sfs/bucket.cc (hunk 1 above)
Comment (irq0) [reply src/rgw/driver/sfs/bucket.cc:463]: try_resolve_mp_from_oid may be unsuccessful if 
- something not LC called without an upload_id (I did not find any place, but that may change)
- LC called, but the MP is already gone

We can't 100% distinguish LC vs non-LC callers. The LC caller passes infos different enough from others, that we can just try to find a LC MP with try_resolve_mp_from_oid. In the LC case oid is a database id - in the non-LC case oid is an object_name. My read of the SAL / Op code is that 'oid' used to refer to a Ceph object id not a S3 object key. We are a bit lost in translation as our IDs are numbers and our names aren't part of the identifier space.

Basically the code does what we did before if we could not resolve to a LC MP

**Review by 0xavi0 — APPROVED** [review]
cool, looks like LC has that corner case covered.

lg2m
